src/nvim/ui_client.c

```diff
@@ -283,6 +283,62 @@ void ui_client_event_raw_line(GridLineEvent *g)
                (const schar_T *)grid_line_buf_char, grid_line_buf_attr);
 }
 
+/// When a "connect" UI event is received, its arguments are saved here when
+/// waiting for the server to exit.
+static Array connect_args;
+static bool connect_pending = false;
+static bool connect_successful = false;
```

Comment:
Now we have both `restart_pending` and `connect_pending`. That's not good. There should be only 1 flag, and if it is set and we handle another event that would set it, that should be an error of some kind.

I also don't think we need separate `connect_args` and `restart_args`.

`connect_successful` also looks suspicious. 

Please think about how we can avoid state where possible, and reduce the number of edge cases and conflicting states.

Reply:
I think this entire postponing isn't needed at all. See my comment below.